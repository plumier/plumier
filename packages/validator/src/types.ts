export type Class<T = any> = new (...args: any[]) => T
