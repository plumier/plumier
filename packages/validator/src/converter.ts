import { Class } from "./types"

// --------------------------------------------------------------------- //
// ------------------------------- TYPES ------------------------------- //
// --------------------------------------------------------------------- //

type Converter = (value: {}) => {} | undefined

// --------------------------------------------------------------------- //
// ------------------------------- HELPER ------------------------------ //
// --------------------------------------------------------------------- //

//some object can't simply convertible to string https://github.com/emberjs/ember.js/issues/14922#issuecomment-278986178
function safeToString(value: any) {
    try {
        return value.toString()
    } catch (e) {
        return "[object Object]"
    }
}


// --------------------------------------------------------------------- //
// ----------------------------- CONVERTERS ---------------------------- //
// --------------------------------------------------------------------- //

function booleanConverter(rawValue: {}) {
    const value = safeToString(rawValue)
    const list: { [key: string]: boolean | undefined } = {
        on: true, true: true, "1": true, yes: true,
        off: false, false: false, "0": false, no: false
    }
    return list[value.toLowerCase()]
}

function numberConverter(rawValue: {}) {
    const value = safeToString(rawValue)
    if (value === "") return
    const result = Number(value)
    if (isNaN(result)) return
    return result
}

function dateConverter(rawValue: {}) {
    const value = safeToString(rawValue)
    const stamp = Date.parse(value)
    return isNaN(stamp) ? undefined : new Date(stamp)
}

function stringConverter(rawValue: {}) {
    return safeToString(rawValue)
}

const defaultConverters = new Map<Class, Converter>([
    [Boolean, booleanConverter],
    [Date, dateConverter],
    [Number, numberConverter],
    [String, stringConverter]
])

export { defaultConverters, Converter, safeToString }
