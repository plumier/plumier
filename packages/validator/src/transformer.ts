import reflect, { useCache } from "@plumier/reflect"

import { Class } from "./types"
import { Converter, defaultConverters } from './converter';

// --------------------------------------------------------------------- //
// ------------------------------- TYPES ------------------------------- //
// --------------------------------------------------------------------- //

export interface PrimitiveNode {
    kind: "Primitive"
    type: Class
    converter: Converter
}

export interface ObjectNode {
    kind: "Object",
    type: Class
    properties: Map<string, SuperNode>
}

export interface ArrayNode {
    kind: "Array",
    type: Class
    element: SuperNode
}

export type SuperNode = PrimitiveNode | ObjectNode | ArrayNode

// --------------------------------------------------------------------- //
// -------------------------------- AST -------------------------------- //
// --------------------------------------------------------------------- //

function primitiveNode(type: Class): PrimitiveNode {
    return {
        kind: "Primitive", type,
        converter: defaultConverters.get(type)!
    }
}

function arrayNode(type: Class[]): ArrayNode {
    let defer: SuperNode
    return {
        kind: "Array", type: type[0],
        get element() {
            return defer || (defer = transform(type[0]))
        }
    }
}

function objectNode(type: Class): ObjectNode {
    let defer: Map<string, SuperNode>
    return {
        kind: "Object", type,
        get properties() {
            if (defer)
                return defer
            else {
                const meta = reflect(type)
                return defer = new Map(meta.properties.map(x => (<[string, ObjectNode]>[x.name, transformer(x.type)])))
            }
        }
    }
}

// --------------------------------------------------------------------- //
// ---------------------------- TRANSFORMER ---------------------------- //
// --------------------------------------------------------------------- //

function transformer(type: Class | Class[]) {
    if (type === Boolean || type === Number || type === Date || type === String)
        return primitiveNode(type)
    else if (Array.isArray(type))
        return arrayNode(type)
    else
        return objectNode(type)
}

// --------------------------------------------------------------------- //
// ------------------------------- CACHE ------------------------------- //
// --------------------------------------------------------------------- //

const astCache = new Map<Class, SuperNode>()
const arrayAstCache = new Map<Class, SuperNode>()

const transformArray = useCache(arrayAstCache, transformer, (type: any) => type[0])
const transformAll = useCache(astCache, transformer, (type) => type)

function transform(type: Class | Class[]) {
    const cachedTransformer = Array.isArray(type) ? transformArray : transformAll
    return cachedTransformer(type)
}

export { transform }