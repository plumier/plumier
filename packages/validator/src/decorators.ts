import { TypeOverride } from "@plumier/reflect"
import validatorJs from "validator"

import * as v from "./validation"

interface Opt { message?: string }

namespace val {
    function check(validator: (x: string) => boolean, message: string) {
        return v.createValidation(x => !validator(x) ? message : undefined)
    }

    export function after(opt?: Opt & { date?: string }) {
        return check(x => validatorJs.isAfter(x, opt && opt.date), opt && opt.message || `Date must be greater than ${opt && opt.date || "today"}`)
    }

    export function alpha(opt?: Opt & { locale?: validatorJs.AlphaLocale }) {
        return check(x => validatorJs.isAlpha(x, opt && opt.locale), opt && opt.message || "Invalid alpha")
    }

    export function alphanumeric(opt?: Opt & { locale?: validatorJs.AlphaLocale }) {
        return check(x => validatorJs.isAlphanumeric(x, opt && opt.locale), opt && opt.message || "Invalid alpha numeric")
    }

    export function ascii(opt?: Opt) {
        return check(x => validatorJs.isAscii(x), opt && opt.message || "Invalid ascii")
    }

    export function base64(opt?: Opt) {
        return check(x => validatorJs.isBase64(x), opt && opt.message || "Invalid base 64")
    }

    export function before(opt?: Opt & { date?: string }) {
        return check(x => validatorJs.isBefore(x, opt && opt.date), opt && opt.message || `Date must be less than ${opt && opt.date || "today"}`)
    }

    export function byteLength(opt: Opt & validatorJs.IsByteLengthOptions) {
        return check(x => validatorJs.isByteLength(x, opt), opt && opt.message || "Invalid byte length")
    }

    export function creditCard(opt?: Opt) {
        return check(x => validatorJs.isCreditCard(x), opt && opt.message || "Invalid credit card number")
    }

    export function currency(opt?: Opt & validatorJs.IsCurrencyOptions) {
        return check(x => validatorJs.isCurrency(x, opt), opt && opt.message || "Invalid currency")
    }

    export function dataURI(opt?: Opt) {
        return check(x => validatorJs.isDataURI(x), opt && opt.message || "Invalid data URI")
    }

    export function decimal(opt?: Opt & validatorJs.IsDecimalOptions) {
        return check(x => validatorJs.isDecimal(x, opt), opt && opt.message || "Invalid decimal")
    }

    export function divisibleBy(opt: Opt & { num: number }) {
        return check(x => validatorJs.isDivisibleBy(x, opt.num), opt && opt.message || `Not divisible by ${opt.num}`)
    }

    export function email(opt?: Opt & validatorJs.IsEmailOptions) {
        return check(x => validatorJs.isEmail(x, opt), opt && opt.message || "Invalid email address")
    }

    export function fqdn(opt?: Opt & validatorJs.IsFQDNOptions) {
        return check(x => validatorJs.isFQDN(x, opt), opt && opt.message || "Invalid FQDN")
    }

    export function float(opt?: Opt & validatorJs.IsFloatOptions) {
        return check(x => validatorJs.isFloat(x, opt), opt && opt.message || "Invalid float number")
    }

    export function fullWidth(opt?: Opt) {
        return check(x => validatorJs.isFullWidth(x), opt && opt.message || "Invalid value provided")
    }

    export function halfWidth(opt?: Opt) {
        return check(x => validatorJs.isHalfWidth(x), opt && opt.message || "Invalid value provided")
    }

    export function hash(opt: Opt & { algorithm: validatorJs.HashAlgorithm }) {
        return check(x => validatorJs.isHash(x, opt.algorithm), opt && opt.message || "Invalid hash")
    }

    export function hexColor(opt?: Opt) {
        return check(x => validatorJs.isHexColor(x), opt && opt.message || "Invalid hex color")
    }

    export function hexadecimal(opt?: Opt) {
        return check(x => validatorJs.isHexadecimal(x), opt && opt.message || "Invalid hexadecimal")
    }

    export function ip(opt?: Opt & { version?: "4" | "6" }) {
        return check(x => validatorJs.isIP(x, opt && opt.version), opt && opt.message || "Invalid IP address")
    }

    export function isbn(opt?: Opt & { version?: "10" | "13" }) {
        return check(x => validatorJs.isISBN(x, opt && opt.version), opt && opt.message || "Invalid ISBN")
    }

    export function isin(opt?: Opt) {
        return check(x => validatorJs.isISIN(x), opt && opt.message || "Invalid ISIN")
    }

    export function iso31661Alpha2(opt?: Opt) {
        return check(x => validatorJs.isISO31661Alpha2(x), opt && opt.message || "Invalid ISO 31661 Alpha 2")
    }

    export function iso8601(opt?: Opt) {
        return check(x => validatorJs.isISO8601(x), opt && opt.message || "Invalid ISO 8601 date")
    }

    export function isrc(opt?: Opt) {
        return check(x => validatorJs.isISRC(x), opt && opt.message || "Invalid ISRC")
    }

    export function issn(opt?: Opt & validatorJs.IsISSNOptions) {
        return check(x => validatorJs.isISSN(x, opt), opt && opt.message || "Invalid ISSN")
    }

    export function int(opt?: Opt & validatorJs.IsIntOptions) {
        return check(x => validatorJs.isInt(x, opt), opt && opt.message || "Invalid integer")
    }

    export function json(opt?: Opt) {
        return check(x => validatorJs.isJSON(x), opt && opt.message || "Invalid JSON")
    }

    export function latLong(opt?: Opt) {
        return check(x => validatorJs.isLatLong(x), opt && opt.message || "Invalid lat long")
    }

    export function length(opt: Opt & validatorJs.IsLengthOptions) {
        return check(x => validatorJs.isLength(x, opt), opt && opt.message || "Invalid length")
    }

    export function lowerCase(opt?: Opt) {
        return check(x => validatorJs.isLowercase(x), opt && opt.message || "Invalid lower case")
    }

    export function macAddress(opt?: Opt) {
        return check(x => validatorJs.isMACAddress(x), opt && opt.message || "Invalid MAC address")
    }

    export function matches(opt: Opt & { pattern: string, modifier?: string }) {
        return check(x => validatorJs.matches(x, opt.pattern, opt.modifier), opt.message || "Invalid string")
    }

    export function md5(opt?: Opt) {
        return check(x => validatorJs.isMD5(x), opt && opt.message || "Invalid MD5 hash")
    }

    export function mimeType(opt?: Opt) {
        return check(x => validatorJs.isMimeType(x), opt && opt.message || "Invalid mime type")
    }

    export function mobilePhone(opt?: Opt & { locale?: validatorJs.MobilePhoneLocale, options?: validatorJs.IsMobilePhoneOptions }) {
        return check(x => validatorJs.isMobilePhone(x, opt && opt.locale, opt && opt.options), opt && opt.message || "Invalid mobile phone")
    }

    export function mongoId(opt?: Opt) {
        return check(x => validatorJs.isMongoId(x), opt && opt.message || "Invalid MongoDb id")
    }

    export function multibyte(opt?: Opt) {
        return check(x => validatorJs.isMultibyte(x), opt && opt.message || "Invalid multi byte")
    }

    export function numeric(opt?: Opt & validatorJs.IsNumericOptions) {
        return check(x => validatorJs.isNumeric(x), opt && opt.message || "Invalid numeric")
    }

    export function port(opt?: Opt) {
        return check(x => validatorJs.isPort(x), opt && opt.message || "Invalid port")
    }

    export function postalCode(opt?: Opt & { locale?: validatorJs.PostalCodeLocale }) {
        return check(x => validatorJs.isPostalCode(x, opt && opt.locale || "any"), opt && opt.message || "Invalid postal code")
    }

    export function slug(opt?: Opt) {
        return check(x => validatorJs.isSlug(x), opt && opt.message || "Invalid slug")
    }

    export function surrogatePair(opt?: Opt) {
        return check(x => validatorJs.isSurrogatePair(x), opt && opt.message || "Invalid surrogate pair")
    }

    export function url(opt?: Opt & validatorJs.IsURLOptions) {
        return check(x => validatorJs.isURL(x, opt), opt && opt.message || "Invalid url")
    }

    export function UUID(opt?: Opt & { version?: 3 | 4 | 5 | "3" | "4" | "5" | "all" }) {
        return check(x => validatorJs.isUUID(x), opt && opt.message || "Invalid UUID")
    }

    export function uppercase(opt?: Opt) {
        return check(x => validatorJs.isUppercase(x), opt && opt.message || "Invalid uppercase")
    }

    export function variableWidth(opt?: Opt) {
        return check(x => validatorJs.isVariableWidth(x), opt && opt.message || "Invalid variable width")
    }

    export function whiteListed(opt: Opt & { chars: string | string[] }) {
        return check(x => validatorJs.isWhitelisted(x, opt && opt.chars), opt && opt.message || "Invalid white listed")
    }

    export function required() {
        return v.required()
    }

    export function partial(typ: TypeOverride | ((x: any) => TypeOverride)) {
        return v.partial(typ)
    }
}


export { val, Opt }