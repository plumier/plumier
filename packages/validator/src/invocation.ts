import { Result, ParentInfo } from "./visitor";
import { SuperNode } from './transformer';
import { Class } from './types';


type VisitorExtension = (next: VisitorInvocation) => Result


interface VisitorInvocation {
    value: {},
    type: Class,
    path: string,
    decorators: any[]
    parent?: ParentInfo
    proceed(): Result
}

class ExtensionInvocationImpl implements VisitorInvocation {
    value: {}
    type: Class
    path: string
    decorators: any[]
    parent?:ParentInfo

    constructor(public ext: VisitorExtension, private next: VisitorInvocation) {
        this.value = next.value
        this.type = next.type
        this.path = next.path
        this.decorators = next.decorators
        this.parent = next.parent
    }

    proceed() {
        return this.ext(this.next)
    }
}

class VisitorInvocationImpl implements VisitorInvocation {
    value: {}
    type: Class
    path: string
    decorators: any[]
    parent?:ParentInfo
    constructor(value: {}, path: string, private ast: SuperNode, decorators: any[], private visitor: () => Result, parent?:ParentInfo) {
        this.value = value
        this.type = ast.type
        this.path = path
        this.decorators = decorators
        this.parent = parent
    }

    proceed(): Result {
        return this.visitor()
    }
}

function pipe(value: {}, strPath: string, ast: SuperNode, decorators: any[], extensions: VisitorExtension[], visitor: () => Result, parent?:ParentInfo) {
    return extensions.reduce((prev, cur) => new ExtensionInvocationImpl(cur, prev),
        <VisitorInvocation>new VisitorInvocationImpl(value, strPath, ast, decorators, visitor, parent))
}

export { pipe, VisitorExtension, VisitorInvocation }