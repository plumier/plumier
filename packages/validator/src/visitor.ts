import reflect, { mergeDecorator } from "@plumier/reflect"

import { safeToString } from "./converter"
import { pipe, VisitorExtension, VisitorInvocation, VisitorInvocationImpl } from "./invocation"
import { ArrayNode, ObjectNode, PrimitiveNode, SuperNode } from "./transformer"
import { Class } from './types';

// --------------------------------------------------------------------- //
// ------------------------------- TYPES ------------------------------- //
// --------------------------------------------------------------------- //

interface ResultMessages {
    path: string,
    messages: string[]
}

interface Result<T = any> {
    value: T,
    issues?: ResultMessages[]
}

interface ParentInfo {
    value: any,
    type: Class
    decorators: any[]
}

interface VisitorOption {
    path: string
    extension: VisitorExtension[]
    decorators: any[],
    guessArrayElement: boolean,
    parent?: ParentInfo
}

namespace Result {
    export function create(value: any): Result {
        return { value }
    }
    export function error(value: any, path: string, message: string | string[]): Result {
        return { value, issues: [{ path, messages: Array.isArray(message) ? message : [message] }] }
    }
}


// --------------------------------------------------------------------- //
// ------------------------------ HELPER ------------------------------- //
// --------------------------------------------------------------------- //

function getPath(path: string, name: string) {
    return path.length === 0 ? name : `${path}.${name}`
}

const visitorMap = {
    "Primitive": primitiveVisitor,
    "Array": arrayVisitor,
    "Object": objectVisitor,
}

function unableToConvert(value: {}, type: string) {
    return `Unable to convert "${safeToString(value)}" into ${type}`
}

// --------------------------------------------------------------------- //
// ------------------------------ VISITORS ----------------------------- //
// --------------------------------------------------------------------- //

function primitiveVisitor(value: {}, ast: PrimitiveNode, opt: VisitorOption): Result {
    const result = ast.converter(value)
    if (result === undefined)
        return Result.error(value, opt.path, unableToConvert(value, ast.type.name))
    else
        return Result.create(result)
}

function arrayVisitor(value: {}[], ast: ArrayNode, opt: VisitorOption): Result {
    const newValues = opt.guessArrayElement && !Array.isArray(value) ? [value] : value
    if (!Array.isArray(newValues)) return Result.error(value, opt.path, unableToConvert(value, `Array<${ast.type.name}>`))
    const result: any[] = []
    const errors: ResultMessages[] = []
    for (let i = 0; i < newValues.length; i++) {
        const val = newValues[i];
        const option = {
            path: getPath(opt.path, i.toString()), extension: opt.extension,
            decorators: opt.decorators, parent: opt.parent,
            guessArrayElement: opt.guessArrayElement
        }
        const elValue = pipeline(val, ast.element as SuperNode, option)
        result.push(elValue.value)
        if (elValue.issues)
            errors.push(...elValue.issues)
    }
    return { value: result, issues: errors.length === 0 ? undefined : errors }
}

function objectVisitor(value: any, ast: ObjectNode, opt: VisitorOption): Result {
    if (typeof value === "number" || typeof value === "string" || typeof value === "boolean")
        return Result.error(value, opt.path, unableToConvert(value, ast.type.name))
    const instance = Object.create(ast.type.prototype)
    const meta = reflect(ast.type)
    const errors: ResultMessages[] = []
    const props = Object.keys(value).reduce((prev, x) => ({ ...prev, [x]: true }), {} as any)
    for (const property of meta.properties) {
        const node = ast.properties.get(property.name) as SuperNode
        // if the property is not provided, then skip
        const option = {
            path: getPath(opt.path, property.name), extension: opt.extension,
            decorators: property.decorators, parent: { value, type: ast.type, decorators: opt.decorators },
            guessArrayElement: opt.guessArrayElement
        }
        const propValue = pipeline(value[property.name], node, option)
        if (propValue.issues)
            errors.push(...propValue.issues)
        if (!props[property.name]) continue
        instance[property.name] = propValue.value
    }
    return { value: instance, issues: errors.length === 0 ? undefined : errors }
}

function visitor(value: any, ast: SuperNode, opt: VisitorOption): Result {
    if (value === undefined || value === null)
        return { value }
    if (value.constructor === ast.type || ast.type === Object)
        return { value: ast.kind === "Array" && opt.guessArrayElement && !Array.isArray(value) ? [value] : value }
    return visitorMap[ast.kind](value, ast as any, opt)
}

function pipeline(value: any, ast: SuperNode, opt: VisitorOption): Result {
    const mainVis = (i: VisitorInvocation) => visitor(i.value, i.ast, opt)
    const last = new VisitorInvocationImpl(mainVis, {
        decorators: opt.decorators, path: opt.path,
        ast: ast, value, parent: opt.parent
    });
    const visitors = pipe(opt.extension, last)
    return visitors.proceed()
}

export { pipeline, ResultMessages, Result, ParentInfo }