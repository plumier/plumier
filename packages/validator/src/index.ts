import { Opt, val } from "./decorators"
import { VisitorExtension, VisitorInvocation } from "./invocation"
import { transform } from "./transformer"
import { Class } from "./types"
import {
    partial,
    PartialValidator,
    Validator,
    validatorVisitor,
    ValidatorDecorator,
    RequiredValidator,
} from "./validation"
import { pipeline, Result, ResultMessages } from "./visitor"
import {defaultConverters} from "./converter"



interface Option<T = any> {
    /**
     * Expected type
     */
    type: T,

    /**
     * List of visitor extension to extend TypedConverter internal process
     */
    visitors?: VisitorExtension[]

    /**
     * Top decorators of the type
     */
    decorators?: any[],

    /**
     * Root path
     */
    path?: string,

    /**
     * Guess single value as array element if expected type is array. Default false
     */
    guessArrayElement?: boolean
}

/**
 * Convert value into type specified on configuration.
 * @param value 
 * @param opt 
 */
function convert<T = any>(value: any, option: Option<T> | T): T extends Class<infer R>[] ? Result<R[]> : T extends Class<infer R> ? Result<R> : Result<any> {
    const opt = "type" in option ? option : { type: option }
    return pipeline(value, transform(opt.type as any), {
        path: opt.path || "", extension: opt.visitors || [],
        decorators: opt.decorators || [],
        guessArrayElement: opt.guessArrayElement || false
    }) as any
}

/**
 * Create type converter with specific configuration
 * @param option 
 */
export default function createConverter<T = any>(option: Option<T> | T) {
    return (value: any) => convert<T>(value, option)
}

function createValidator<T>(option: T | Option<T>) {
    return (value: any) => validate<T>(value, option)
}

function validate<T>(value: any, option: Option<T> | T) {
    const opt = "type" in option ? option : { type: option }
    const visitors = [validatorVisitor]
    for (const visitor of (opt.visitors || [])) {
        visitors.push(visitor)
    }
    return convert(value, {
        decorators: opt.decorators, guessArrayElement: opt.guessArrayElement,
        path: opt.path, type: opt.type, visitors
    })
}

export {
    convert, Option, VisitorExtension,
    VisitorInvocation, Result, ResultMessages,
    partial, validatorVisitor, Validator,
    PartialValidator, val, Opt, validate, createValidator, ValidatorDecorator,
    RequiredValidator, defaultConverters
}