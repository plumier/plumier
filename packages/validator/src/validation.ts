import reflect, { decorateProperty, decorate, mergeDecorator, TypeOverride } from "@plumier/reflect";
import { VisitorInvocation } from './invocation';
import { Result } from './visitor';
import { Class } from './types';
import { safeToString } from './converter';

const RequiredValidator = Symbol("tc:required")
const PartialValidator = Symbol("tc:partial")
type Validator = (val: any) => string | undefined
interface ValidatorDecorator { type: "tc:validator", validator: Validator | string | symbol }

// --------------------------------------------------------------------- //
// ----------------------------- DECORATORS ---------------------------- //
// --------------------------------------------------------------------- //

function createValidation(validator: Validator | string | symbol) {
    return decorateProperty(<ValidatorDecorator>{ type: "tc:validator", validator })
}

function required() {
    return createValidation(RequiredValidator)
}

function partial(type: TypeOverride | ((x: any) => TypeOverride)) {
    return mergeDecorator(reflect.type(type), createValidation(PartialValidator))
}

// --------------------------------------------------------------------- //
// ------------------------------ HELPERS ------------------------------ //
// --------------------------------------------------------------------- //

function isPartial(i: VisitorInvocation) {
    return i.parent && i.parent.decorators.find((x: ValidatorDecorator) => x.validator === PartialValidator)
}

function isEmpty(result: Result) {
    return (result.value === "" || result.value === undefined || result.value === null)
}

// --------------------------------------------------------------------- //
// ------------------------------ VISITORS ----------------------------- //
// --------------------------------------------------------------------- //

function getValidators(i: VisitorInvocation): ValidatorDecorator[] {
    return i.decorators
        .filter((x: ValidatorDecorator) => x.validator !== PartialValidator && x.type === "tc:validator")
}

function validatorVisitor(i: VisitorInvocation): Result {
    const validators = getValidators(i)
    const result = i.proceed()
    if (!isPartial(i) && isEmpty(result) && validators.some(x => x.validator === RequiredValidator))
        return Result.error(i.value, i.path, "Required")
    if (isEmpty(result) || validators.length === 0) return result
    else {
        const messages: string[] = []
        for (const validator of validators) {
            if (typeof validator.validator === "function") {
                const msg = validator.validator(safeToString(i.value))
                if (msg) messages.push(msg)
            }
        }
        return messages.length > 0 ? Result.error(i.value, i.path, messages) : result
    }
}

export {
    createValidation, required, partial,
    validatorVisitor, Validator, ValidatorDecorator,
    PartialValidator, RequiredValidator
}
