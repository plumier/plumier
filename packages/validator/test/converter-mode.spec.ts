import reflect from "@plumier/reflect"

import { convert } from "../src"


describe("Model Converter", () => {
    it("Should convert model and appropriate properties", () => {
        @reflect.parameterProperties()
        class AnimalClass {
            constructor(
                public id: number,
                public name: string,
                public deceased: boolean,
                public birthday: Date
            ) { }
        }

        const result = convert({ id: "200", name: "Mimi", deceased: "ON", birthday: "2018-2-2" }, { type: AnimalClass })
        expect(result.value).toBeInstanceOf(AnimalClass)
        expect(result.value).toEqual({ birthday: new Date("2018-2-2"), deceased: true, id: 200, name: "Mimi" })
    })

    it("Should not convert excess properties", () => {
        @reflect.parameterProperties()
        class AnimalClass {
            constructor(
                public id: number,
                public name: string,
            ) { }
        }

        const result = convert({ id: "200", name: "Mimi", deceased: "ON", birthday: "2018-2-2", }, { type: AnimalClass })
        expect(result.value).toBeInstanceOf(AnimalClass)
        expect(result.value).toEqual({ id: 200, name: "Mimi" })
    })


    it("Should allow undefined value", () => {
        @reflect.parameterProperties()
        class AnimalClass {
            constructor(
                public id: number,
                public name: string,
                public deceased: boolean,
                public birthday: Date
            ) { }
        }

        const result = convert({}, { type: AnimalClass })
        expect(result.value).toBeInstanceOf(AnimalClass)
        expect(result.value).toEqual({})
    })

    it("Should allow undefined value when provided", () => {
        @reflect.parameterProperties()
        class AnimalClass {
            constructor(
                public id: number,
                public name: string,
                public deceased: boolean,
                public birthday: Date
            ) { }
        }

        const result = convert({ id: undefined }, { type: AnimalClass })
        expect(result.value).toBeInstanceOf(AnimalClass)
        expect(result.value).toMatchSnapshot()
    })

    it("Should throw if provided non convertible value", () => {
        @reflect.parameterProperties()
        class AnimalClass {
            constructor(
                public id: number,
                public name: string,
                public deceased: boolean,
                public birthday: Date
            ) { }
        }

        const result = convert({ id: "200", name: "Mimi", deceased: "Hello", birthday: "2018-2-2" }, { type: AnimalClass })
        expect(result.issues).toEqual([{ path: "deceased", messages: [`Unable to convert "Hello" into Boolean`] }])
    })

    it("Should throw if provided non convertible value", () => {
        @reflect.parameterProperties()
        class AnimalClass {
            constructor(
                public id: number,
                public name: string,
                public deceased: boolean,
                public birthday: Date
            ) { }
        }

        const result = convert("Hello", { type: AnimalClass })
        expect(result.issues).toEqual([{ path: "", messages: [`Unable to convert "Hello" into AnimalClass`] }])
    })

    it("Should not populate optional properties with undefined", () => {
        @reflect.parameterProperties()
        class AnimalClass {
            constructor(
                public id: number,
                public name: string,
                public deceased: boolean | undefined,
                public birthday: Date | undefined
            ) { }
        }

        const result = convert({ id: "200", name: "Mimi", excess: "Hola" }, { type: AnimalClass })
        expect(result.value).toBeInstanceOf(AnimalClass)
        expect(Object.keys(result.value)).toEqual(["id", "name"])
        expect(result.value).toEqual({ id: 200, name: "Mimi" })
    })

    it("Should not convert if expected type is Object", () => {
        const result = convert({
            host: '127.0.0.1:61945',
            'accept-encoding': 'gzip, deflate',
            'user-agent': 'node-superagent/3.8.3',
            'content-type': 'application/json',
            'content-length': '67',
            connection: 'close'
        }, { type: Object })
        expect(result.value).toEqual({
            host: '127.0.0.1:61945',
            'accept-encoding': 'gzip, deflate',
            'user-agent': 'node-superagent/3.8.3',
            'content-type': 'application/json',
            'content-length': '67',
            connection: 'close'
        })
    })
})