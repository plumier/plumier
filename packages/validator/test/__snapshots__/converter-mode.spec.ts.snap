// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Model Converter Should allow undefined value when provided 1`] = `
AnimalClass {
  "id": undefined,
}
`;
